File: ch4/traversal/toposort.py
import sys
from enum import Enum

class flag(Enum):
  UNVISITED = -1
  VISITED = -2

AL = []
dfs_num = []
ts = []

def toposort(u):
  global AL
  global dfs_num
  global ts

  dfs_num[u] = flag.VISITED.value
  for v, w in AL[u]:
    if dfs_num[v] == flag.UNVISITED.value:
      toposort(v)
  ts.append(u)


def main():
  global AL
  global dfs_num
  global ts

  fp = open('toposort_in.txt', 'r')

  V = int(fp.readline().strip())
  AL = [[] for _ in range(V)]
  for u in range(V):
    tkn = list(map(int, fp.readline().strip().split()))
    k = tkn[0]
    for i in range(k):
      v, w = tkn[2*i+1], tkn[2*i+2]
      AL[u].append((v, w))

  print('Topological Sort (the input graph must be DAG)')
  dfs_num = [flag.UNVISITED.value] * V
  for u in range(V):
    if dfs_num[u] == flag.UNVISITED.value:
      toposort(u)
  ts = ts[::-1]
  print(' '.join(map(str, ts)))


main()

File: ch4/traversal/toposort.ml
open Scanf
open Printf

type graph = int list array (* Adjacency List representation *)

(** Return a topological ordering of the DAG *)
let topo_sort (adj : graph) : int list =
  let n = Array.length adj in
  let visited = Array.make n false in
  let rec visit order u =
    visited.(u) <- true;
    u ::
      (adj.(u) |> List.fold_left (fun order v ->
        if visited.(v) then order
        else visit order v
      ) order)
  in
  List.init n (fun x -> x) |> List.fold_left (fun acc u ->
    if visited.(u) then acc
    else visit acc u 
  ) []

let () =
  let ic = Scanning.open_in "toposort_in.txt" in
  let sc = object method nf = bscanf ic end in
  let n = sc#nf " %d " (fun x -> x) in
  let adj = Array.init n (fun _ ->
    sc#nf " %d " (fun degree ->
      List.init degree (fun _ -> sc#nf " %d %d " (fun v w -> v))) (* Ignoring weight *)
  ) in
  printf "Topological Sort (the input graph must be DAG)\n";
  topo_sort adj |> List.iter (printf " %d"); printf "\n";
